### PR TITLE
Add FreeSemilattice constructor

### PR DESCRIPTION
**Pull request by Jun2M**
This pull request adds a new constructor for the FreeSemilattice class. The constructor takes a positive integer as input and creates a FreeSemilattice object based on the specified semigroup type. The constructor supports various semigroup types such as IsFpSemigroup, IsTransformationSemigroup, and IsPartialPermSemigroup. This addition enhances the functionality of the FreeSemilattice class and provides more flexibility in creating semilattices.

### CODE REVIEW

File: doc/semicons.xml [hunk 1]
```diff
@@ -147,6 +147,47 @@ true]]></Example>
   </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="FreeSemilattice">
+  <ManSection>
+    <Func Name="FreeSemilattice" Arg="[filt, ] n"/>
+    <Returns>
+      A free semilattice of <A>n</A> generators.
```
Comment (james-d-mitchell) [inline doc/semicons.xml:154]:
```suggestion
      A free semilattice with <A>n</A> generators.
```

File: gap/semigroups/semicons.gi [hunk 2]
```diff
@@ -406,6 +406,117 @@ for _IsXSemigroup in ["IsBooleanMatSemigroup",
...
+  local filter, n, S;
+
+  if Length(arg) = 1  then
+    filter := IsTransformationSemigroup;
+    n := arg[1];
+  elif Length(arg) = 2 then
+    filter := arg[1];
+    n := arg[2];
+  else
+    ErrorNoReturn("the arguments must be a positive integer or a filter ",
```
Comment (james-d-mitchell) [inline gap/semigroups/semicons.gi:422]:
```suggestion
    ErrorNoReturn("expected 2 arguments found ", Length(arg));
```

File: gap/semigroups/semicons.gi [hunk 3]
```diff
@@ -406,6 +406,117 @@ for _IsXSemigroup in ["IsBooleanMatSemigroup",
...
+                  "and a positive integer");
```
Comment (james-d-mitchell) [inline gap/semigroups/semicons.gi:423]:
```suggestion
```

File: gap/semigroups/semicons.gi [hunk 4]
```diff
@@ -406,6 +406,117 @@ for _IsXSemigroup in ["IsBooleanMatSemigroup",
...
+  # SetIsRectangularBand(S, true);
+  # SetNrRClasses(S, m);
+  # SetNrLClasses(S, n);
+  # if m <> 1 or n <> 1 then
+  #   SetIsGroupAsSemigroup(S, false);
+  #   SetIsZeroSemigroup(S, false);
+  #   SetIsTrivial(S, false);
+  # fi;
+  # SetIsRightZeroSemigroup(S, m = 1);
+  # SetIsLeftZeroSemigroup(S, n = 1);
```
Comment (james-d-mitchell) [inline gap/semigroups/semicons.gi:443]:
```suggestion
```

**Review by james-d-mitchell — APPROVED** [review]
I think the only question I have is should the returned semilattice have `2^n` elements or `(2^n) - 1` elements as you've implemented?  @Jun2M can you possibly say why you chose this approach?

Otherwise this looks great!

File: gap/semigroups/semicons.gi [hunk 5]
```diff
@@ -406,6 +406,106 @@ for _IsXSemigroup in ["IsBooleanMatSemigroup",
...
+  fi;
+
+  if not IsPosInt(n) or not IsOperation(filter) then
+    ErrorNoReturn("the arguments must be a positive integer or a filter ",
+                  "and a positive integer");
+  fi;
+
+  S := FreeSemilatticeCons(filter, n);
+
+  SetSize(S, 2 ^ n - 1);
```
Comment (james-d-mitchell) [inline gap/semigroups/semicons.gi:432]: Probably also `SetIsSemilattice` to true also

**Comment by james-d-mitchell** [comment]
@Jun2M I don't think the CI tests failing are anything to do the changes in your PR, there seems to be an issue with GAP at the moment:

https://github.com/gap-system/gap/issues/5672

**Comment by james-d-mitchell** [comment]
CI re-run after https://github.com/gap-system/gap/issues/5672 closed. Fingers crossed...

**Comment by Jun2M** [comment]
The check was not successful because the test for NumberOfLeftCongruences failed? I don't think I touched that code at all..?

**Comment by james-d-mitchell** [comment]
Probably just a spurious error, running it again :)

**Pull request merged**